lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+ *
+ * @param data object keyed by ONYXKEYS and the values to set
+ */
+declare function multiSet(data: Partial<KeyValueMap>): Promise<void>;
+
+/**
+ * Merge a new value into an existing value at a key.
+ *
+ * The types of values that can be merged are `Object` and `Array`. To set another type of value use `Onyx.set()`. Merge
+ * behavior uses lodash/merge under the hood for `Object` and simple concatenation for `Array`. However, it's important
+ * to note that if you have an array value property on an `Object` that the default behavior of lodash/merge is not to
+ * concatenate. See here: https://github.com/lodash/lodash/issues/2872
+ *
+ * Calls to `Onyx.merge()` are batched so that any calls performed in a single tick will stack in a queue and get
+ * applied in the order they were called. Note: `Onyx.set()` calls do not work this way so use caution when mixing
+ * `Onyx.merge()` and `Onyx.set()`.
+ *
+ * @example
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Joe']); // -> ['Joe']
+ * Onyx.merge(ONYXKEYS.EMPLOYEE_LIST, ['Jack']); // -> ['Joe', 'Jack']
+ * Onyx.merge(ONYXKEYS.POLICY, {id: 1}); // -> {id: 1}
+ * Onyx.merge(ONYXKEYS.POLICY, {name: 'My Workspace'}); // -> {id: 1, name: 'My Workspace'}
+ *
+ * @param key ONYXKEYS key
+ * @param value Object or Array value to merge
+ */
+declare function merge<TKey extends Key | CollectionKey>(key: TKey, value: PartialDeep<Value[TKey]>): Promise<void>;
+
+/**
+ * Clear out all the data in the store
+ *
+ * Note that calling Onyx.clear() and then Onyx.set() on a key with a default
+ * key state may store an unexpected value in Storage.
+ *
+ * E.g.
+ * Onyx.clear();
+ * Onyx.set(ONYXKEYS.DEFAULT_KEY, 'default');
+ * Storage.getItem(ONYXKEYS.DEFAULT_KEY)
+ *     .then((storedValue) => console.log(storedValue));
+ * null is logged instead of the expected 'default'
+ *
+ * Onyx.set() might call Storage.setItem() before Onyx.clear() calls
+ * Storage.setItem(). Use Onyx.merge() instead if possible. Onyx.merge() calls
+ * Onyx.get(key) before calling Storage.setItem() via Onyx.set().
+ * Storage.setItem() from Onyx.clear() will have already finished and the merged
+ * value will be saved to storage after the default value.
+ *
+ * @param keysToPreserve is a list of ONYXKEYS that should not be cleared with the rest of the data
+ */
+declare function clear(keysToPreserve?: (Key | CollectionKey)[]): Promise<void>;
+
+/**
+ * Merges a collection based on their keys
+ *
+ * @example
+ *
+ * Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, {
+ *     [`${ONYXKEYS.COLLECTION.REPORT}1`]: report1,
+ *     [`${ONYXKEYS.COLLECTION.REPORT}2`]: report2,
+ * });
+ *
+ * @param collectionKey e.g. `ONYXKEYS.COLLECTION.REPORT`
+ * @param collection Object collection keyed by individual collection member keys and values
+ */
+declare function mergeCollection<TKey extends CollectionKey, TMap>(collectionKey: TKey, collection: MergeCollection<TKey, TMap, PartialDeep<Value[TKey]>>): Promise<void>;
```

Comment:
Can the signature of `mergeCollection` be simplified like below?

```suggestion
declare function mergeCollection<TKey extends CollectionKey>(collectionKey: TKey, collection: Record<`${TKey}${string}`, PartialDeep<Value[TKey]>>): Promise<void>;
```

I know this looses the type checking to make sure collection id is post fixed, but this seems much simpler.

Reply:
one more question. with the current type signature, do you have to pass `TMap` type argument?

Reply:
> one more question. with the current type signature, do you have to pass TMap type argument?

Yes, it's required so that this magic type works 😅

Reply:
> Can the signature of mergeCollection be simplified like below?

It is a trade-off, we can either make the type more readable or strict. With the typings that @hayata-suenaga suggested I believe the type is compromised too much:

```ts
type Key = 'download_' | 'report_';
declare function mergeCollection<K extends Key>(collectionKey: K, collection: Record<`${K}${string}`, unknown>): Promise<void>;

mergeCollection('download_', { 
    download_id: 'test', // OK
    download_test: 'test', // OK

    // report_test: 'test', // Errors 

    [`${'report_'}id`]: 'test', // Doesn't error
    [`${'report'}id`]: 'test', // Doesn't error
    [`${'download_'}`]: 'test', // Doesn't error
    [`${'whatever'}`]: 'test', // Doesn't error
})
```

The problem is that template literal strings are inferred as `string` and we lose type safety (almost everything can be passed as an argument)

Reply:
yea but we have to pass `as const` anyways, right just as @fabioh8010 did [in his examples](https://github.com/Expensify/react-native-onyx/pull/267#:~:text=Onyx.mergeCollection())?

And in the example, `TMap` is omitted but it's required right?

I'd like to know how much type safety we can get by using `MergeCollection` type and manually passing `TMap` for every `mergeCollection` invocation